src/crl/interactivesessions/shells/shell.py

```diff
@@ -282,11 +282,11 @@ def _send_input(self, _input):
         if self._tty_echo:
             self._read(len(_input))
 
-    def _send_input_line(self, _input):
+    def _send_input_line(self, _input, remove_from_buffer=True):
         """Send a line of input to the terminal."""
         self._terminal.sendline(_input)
 
-        if self._tty_echo:
+        if self._tty_echo and remove_from_buffer:
```

Comment:
This removal is for echo of the sendline command. This business logic violates the original idea of the method. The idea is to send command to the terminal. If the terminal echo is on, then the echo of the command is first written by the slave pty. 

So, I guess there is a problem in the terminal echo not written by slave. This is quite same as saying that the terminal echo is not on. Therefore, it would be better to turn the terminal echo off by from the Shell via setting self._tty_echo=False.